sdk/program-entrypoint/src/lib.rs

```diff
@@ -279,6 +279,85 @@ macro_rules! custom_panic_default {
     };
 }
 
+/// This is a more space efficient implementation of custom panic.
+/// It is about 1kb in size and depends on the unstable feature `panic_info_message`,
+/// which is going to be stabilized in Rust 1.84.
+///
+/// It works just like the default custom panic, except that dynamic generated error
+/// messages are not shown. For instance, trying to access an index out of bounds of a vector
+/// would give us `index Y is out of bounds for length X`. As the length is only know at runtime
+/// time, this message is elided.
+///
+/// All messages known at compile time are correctly displayed, e.g. `called unwrap in a None`,
+/// file names, line and column numbers.
+#[macro_export]
+macro_rules! custom_panic_space_efficient {
+    () => {
+        #[cfg(all(not(feature = "custom-panic"), target_os = "solana"))]
+        #[no_mangle]
+        fn custom_panic(info: &core::panic::PanicInfo<'_>) {
+            #[inline(never)]
+            unsafe fn write_num(mut num: u32, dst: *mut u8) -> usize {
+                let mut buf: [u8; 10] = [0; 10];
+                let init_ptr = buf.as_mut_ptr().add(11);
+                let mut write_ptr = init_ptr;
+                while num > 0 {
+                    write_ptr = write_ptr.sub(1);
+                    // SAFETY: The pointer will always be within *buf and *buf+10
+                    *write_ptr = (num % 10) as u8 + 48;
+                    num /= 10;
+                }
+                *dst = 58;
+                let dst = dst.add(1);
+                let len = init_ptr.offset_from(write_ptr) as usize;
+                // SAFETY: The copy length will never be greater than 10.
+                std::ptr::copy_nonoverlapping(write_ptr, dst, len);
+                len + 1
+            }
+
+            if let Some(loc) = info.location() {
+                let filename = loc.file().as_bytes();
+                // Vector of filename size + 11 (line number + separator)
+                // + 11 (column number + separator) + 13 (panic string) + 15 extra bytes
+                let mut msg_line = vec![32; filename.len() + 50];
```

Comment:
Could we have a "large" fixed length buffer instead of a `Vec`? Worst case we could "truncate" the filename if it goes over a pre-defined limit. This would be a "hybrid" solution between this PR and #3952. 😅

Reply:
I have thought of truncating the file name, but that comes with some problems. First, we would need to choose a length to clip the name. There is no magic number, so we would risk printing gibberish or pointing to the wrong file.

Second, we may include a simple algorithm to clip the file path exactly at a slash, but that would ultimately make `custom_panic` longer.